sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -567,8 +572,13 @@ template <typename T, int N, typename AccessorTy>
 __ESIMD_API std::enable_if_t<(sizeof(T) <= 4) &&
                              (N == 1 || N == 8 || N == 16 || N == 32) &&
                              !std::is_pointer<AccessorTy>::value>
-scatter(AccessorTy acc, simd<uint32_t, N> offsets, simd<T, N> vals,
-        uint32_t glob_offset = 0, simd_mask<N> mask = 1) {
+scatter(AccessorTy acc,
+#ifdef __ESIMD_FORCE_STATELESS_MEM
```

Comment:
Maybe it was a previous mistake having simd<Toffset,N> type for 'offsets' here: https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/esimd/memory.hpp#L209

If conversion to vector of uint64_t happens anyway, then why not simplify that prototype and specify it there as:
```
template <typename Tx, int N>
__ESIMD_API simd<Tx, N> gather(const Tx *p, simd<uint64_t, N> offsets,
                               simd_mask<N> mask = 1) {
```
and then remove the line https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/esimd/memory.hpp#L136

Reply:
Gregory, do you mean by this: "_It seems simd just silently converts data so you don't actually impose any conditions on offset type_" that 64-bit offset passed to stateful gather is silently truncated to 32-bit offset, which doesn't happen for stateless gather?

The approach used here is consistent with previous PR: https://github.com/intel/llvm/pull/9160
AFAIR, we discussed this during teams meeting and it did not meet objections. Do you think it is a mistake in 9160 not reporting an static_error on attempt to pass uint64_t offset to stateful block load?

Reply:
if we look at the CI failure, we see we actually rely on the truncating today:

https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/ESIMD/accessor_gather_scatter.hpp#L37

the result of
` offsets * sizeof(T)` is simd<unsigned long, N>`

so my change causes an error when it used to work.

maybe we should just update the test, but maybe users are relying on this behavior working and we should not error?

Reply:
>As far as I remember the reason why it was done in the previous PR was to preserve some warning we were emitting when passing 64 bit offsets to stateful version.

Yes, that's right.

Here there are two items to look at:
1) The global offset

In the current version of the PR, we preserve the warning in stateful mode with 64-bit offset because we use an `ifdef` for param type to be 32-bit

2) The per-element offset

In the current version of the PR, we do not error or warn. This is because conversion happened anyway (32->64 and 64->32) in both stateful and stateless today, and we have at least one test that does this today (in stateful mode), so my intuition is that users may be doing it too.